src/Pim/Bundle/CatalogBundle/Doctrine/ORM/CompletenessRemover.php

```diff
@@ -123,17 +141,46 @@ protected function bulkRemoveCompletenesses(CursorInterface $products)
                 $bulkCounter++;
             }
 
-            $product->getCompletenesses()->clear();
+            $this->clearProductCompleteness($product, $channel, $locale);
         }
 
         if (!empty($productIds)) {
-            $statement->bindValue('productIds', $productIds, Type::SIMPLE_ARRAY);
-            $statement->execute();
+            $this->entityManager->getConnection()->executeQuery(
+                $query,
+                array_merge([$productIds], $queryParams),
+                $queryTypes
+            );
             $this->indexer->indexAll($bulkedProducts);
         }
     }
 
     /**
+     * @param ProductInterface      $product
+     * @param null|ChannelInterface $channel
+     * @param null|LocaleInterface  $locale
+     */
+    protected function clearProductCompleteness(
+        ProductInterface $product,
+        ChannelInterface $channel = null,
+        LocaleInterface $locale = null
+    ) {
+        if (null === $channel && null === $locale) {
+            $product->getCompletenesses()->clear();
+        } else {
+            $productCompletenesses = $product->getCompletenesses();
+            $completenessesToRemove = $productCompletenesses->filter(
```

Comment:
Alright I get all this functional stuff, but here you could do one for each and one if then remove completeness. Don't revert this ^^ It's just I find the code harder to understand that way ^^